node/src/components/consensus/era_supervisor.rs

```diff
@@ -685,14 +719,18 @@ where
                             .cloned()
                             .expect("switch block should have era validator weights"),
                         switch_block.height() + 1,
-                        *switch_block.state_root_hash(),
+                        *switch_block.state_root_hash(), /* TODO: wrong if we want to cater to
+                                                          * upgrades modifying global state */
                         switch_block.timestamp(),
                     )
                 } else {
-                    // TODO: the start height should be 1 more than the last block before the last
-                    // upgrade, and the state hash should be the (potentially modified) state hash
-                    // from that block
-                    (validators.clone(), 0, state_root_hash, genesis_start_time)
+                    // If we don't have a switch block, we're dealing with era 0 and genesis data
+                    (
+                        genesis_validators.clone(),
+                        0,
+                        genesis_state_root_hash,
+                        genesis_start_time,
+                    )
                 };
```

Comment:
Can we extract this whole `if { … } else { … }` to a method and create a newtype struct for the returned arguments?

Reply:
I'm not sure how you would see it? Some small struct with these 4 fields and a method taking a switch block and returning either the fields, or the values from the switch block?

Reply:
In both branches of the `if-else` we return `(validators, init block height, init state root hash, era start timestamp)`. This can be wrapped in a newtype struct.